### PR TITLE
CSHARP-4090 / CSHARP-4092

### PR DESCRIPTION
**Pull request by JamesKovacs**
CSHARP-4090: Implement support for string.Contains(match, StringComparison).
CSHARP-4092: Fix case-insensitive StartsWith/EndsWith to generate correct MQL.

### CODE REVIEW

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/StringMethod.cs [hunk 1]
```diff
@@ -24,6 +24,9 @@ internal static class StringMethod
     {
         // private static fields
         private static readonly MethodInfo __contains;
+#if NETSTANDARD2_1
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/StringMethod.cs:27]: We've never used it before but I'm thinking `NETSTANDARD2_1_OR_GREATER` might be the appropriate conditional symbol here.

Comment (rstam) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/StringMethod.cs:27]: And if so similarly in other files changed in this PR.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/StartsWithContainsOrEndsWithMethodToAggregationExpressionTranslator.cs [hunk 2]
```diff
@@ -86,7 +90,7 @@ public static AggregationExpression Translate(TranslationContext context, Method
                 if (ignoreCase)
                 {
                     stringAst = AstExpression.ToLower(stringAst);
-                    substringAst = AstExpression.ToLower(stringAst);
+                    substringAst = AstExpression.ToLower(substringAst);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToAggregationExpressionTranslators/MethodTranslators/StartsWithContainsOrEndsWithMethodToAggregationExpressionTranslator.cs:93]: Good catch.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        public void Should_throw_not_supported_exception_for_starts_with_with_unsupported_string_comparison_type(StringComparison comparison)
+        {
+            var collection = GetCollection<C>();
+
+            var exception = Record.Exception(() => collection.Find(x => x.Text.StartsWith("orange", comparison)).ToList());
+
+            exception.Should().BeOfType<ExpressionNotSupportedException>();
+        }
+
+#if NETCOREAPP3_1
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs:64]: If we decide to use `NETSTANDARD2_1_OR_GREATER` in the driver then let's use `NETCOREAPP3_1_OR_GREATER` in the test project.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            var collection = GetCollection<C>();
+
+            var exception = Record.Exception(() => collection.Find(x => x.Text.EndsWith("orange", comparison)).ToList());
+
+            exception.Should().BeOfType<ExpressionNotSupportedException>();
+        }
+
+        public class C
+        {
+            public ObjectId Id { get; set; }
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs:144]: Unless it is important to the test for the `_id` to be an `ObjectId` it's always easier to use `int` for the `_id`.

**Review by rstam — COMMENTED** [review]
I'm willing to LGTM as is but I'm holding off until you read the comments and respond.

**Comment by rstam** [comment]
The way we use `StringComparison.CurrentCulture` in this PR is consistent with how it was already being used by the existing support for `StartsWith` and `EndsWith`, so that's good.

We might still need to review *other* uses of `StringComparison` to make sure we are consistent. In particular that "current culture" maps to "pipeline culture" and whether "ordinal" or "invariant" are supported at all.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/StringMethod.cs (hunk 1 above)
Comment (JamesKovacs) [reply src/MongoDB.Driver/Linq/Linq3Implementation/Reflection/StringMethod.cs:27]: Valid point. Updated in all files.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs (hunk 3 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs:64]: Done.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs (hunk 4 above)
Comment (JamesKovacs) [reply tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4090Tests.cs:144]: Updated.

**Review by JamesKovacs — COMMENTED** [review]
Responded to feedback.

**Comment by JamesKovacs** [comment]
> We might still need to review _other_ uses of `StringComparison` to make sure we are consistent. In particular that "current culture" maps to "pipeline culture" and whether "ordinal" or "invariant" are supported at all.

`EqualsMethodToAggregationExpressionTranslator` considers `CurrentCulture` and `Ordinal` to be equivalent. There may be other places. I'll create a separate ticket to review our usages of `CurrentCulture` and `Ordinal` across the driver. Created https://jira.mongodb.org/browse/CSHARP-4100 to capture this work.

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**